File: DataStructure/WordSearchII212.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>

using namespace std;

class TrieNode {
public:
    unordered_map<char, TrieNode *> chars;
    string word = "";
    ~TrieNode(){
        for (auto & p: chars) delete p.second;
    }
};


class Solution {
    TrieNode * root;
    vector<string> res;

    void insertWord (string word) {
        TrieNode * cur = root;
        for (char c: word){
            if (cur->chars.find(c) == cur->chars.end()){
                cur->chars[c] = new TrieNode();
            }
            cur = cur->chars[c];
        }
        cur->word = word;
    }

    void dfs(int i, int j, TrieNode* cur, vector<vector<char>>& board){
        if (board[i][j] == '*' || cur->chars.find(board[i][j]) == cur->chars.end()) return;
        char ch = board[i][j];
        cur = cur->chars[ch];
        if (!cur->word.empty()) {
            res.push_back(cur->word);
            cur->word = "";  // avoid duplicate word being added to result
        }
        board[i][j] = '*';
        if (i > 0) dfs(i-1, j, cur, board);
        if (j > 0) dfs(i, j - 1, cur, board);
        if (i < board.size() - 1) dfs(i + 1, j, cur, board);
        if (j < board[0].size() - 1) dfs(i, j + 1, cur, board);
        board[i][j] = ch; // reset
    }

public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        root = new TrieNode();
        res.clear();
        for (auto & word: words) insertWord(word); // build trie
        for (int i = 0; i < board.size(); ++i){
            for (int j = 0; j < board[0].size(); ++j){
                dfs(i, j, root, board);
            }
        }
        return res;
    }

    ~Solution(){
        delete root;
        res.clear();
    }
};

File: DataStructure/WordSearchII212.java
import java.util.List;
import java.util.HashMap;
import java.util.ArrayList;

class TrieNode {
    HashMap<Character, TrieNode> chars = new HashMap<>();
    String word = null;
}

public class WordSearchII212 {
    private TrieNode root;
    private List<String> res;
    private void insertWord (String word) {
        TrieNode cur = root;
        for (char c: word.toCharArray()){
            if (!cur.chars.containsKey(c)){
                cur.chars.put(c, new TrieNode());
            }
            cur = cur.chars.get(c);
        }
        cur.word = word;
    }
    public List<String> findWords(char[][] board, String[] words) {
        root = new TrieNode();
        res = new ArrayList<>();
        // Build trie
        for (String word: words) {
            insertWord(word);
        }
        int m = board.length;
        int n = board[0].length;
        for (int i = 0; i < m; i++){
            for (int j = 0; j < n; j++){
                dfs(i, j, root, board);
            }
        }
        return res;
    }

    public void dfs(int i, int j, TrieNode cur, char[][] board) {
        char ch = board[i][j];
        if (ch != '*' && cur.chars.containsKey(ch)) {
            cur = cur.chars.get(ch);
            if (cur.word != null) {
                res.add(cur.word);
                cur.word = null; // avoid duplicate word being added to result
            }
            board[i][j] = '*'; // already visited
            if (i > 0) dfs(i-1, j, cur, board);
            if (j > 0) dfs(i, j-1, cur, board);
            if (i < board.length - 1) dfs(i+1, j, cur, board);
            if (j < board[0].length - 1) dfs(i, j+1, cur, board);
            board[i][j] = ch;
        }
    }
}
